code/datums/components/machine_power_modifier.dm

```diff
@@ -0,0 +1,30 @@
+/datum/component/machine_power_modifier
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+	var/idle_power_multiplier = 1
+	var/active_power_multiplier = 1
+
+/datum/component/machine_power_modifier/Initialize(idle_power_multiplier, active_power_multiplier)
+	if(!istype(parent, /obj/machinery) || istype(parent, /obj/machinery/power) || istype(parent, /obj/machinery/computer))
+		return COMPONENT_INCOMPATIBLE
+	src.idle_power_multiplier = idle_power_multiplier
+	src.active_power_multiplier = active_power_multiplier
+	RegisterSignal(parent, COMSIG_MACHINERY_REFRESH_PARTS, .proc/on_refresh_parts)
+	RegisterSignal(parent, COMSIG_OBJ_DECONSTRUCT, .proc/on_machine_deconstruct)
+	ADD_TRAIT(parent, TRAIT_MACHINE_POWER_UPGRADED, src)
```

Comment:
To further generalize this, the trait you apply should also be passed into the component as an optional parameter.